### PR TITLE
Remove Meshkind from backward, move functionality to evaluate_mesh

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

Method backward now only works on quadrature points, as it should. It is now just the inverse of forward, which is also just using quadrature points.

Move the possibility of evaluating a spectral function on a uniform mesh to evaluate_mesh.

## Types of Changes

- [ ] Bug fix
- [x] New feature
- [x] Performance improvement
- [x] Documentation update
- [x] Refactor / cleanup
- [x] Build / CI
- [ ] Other (describe):

## Description of Changes

* Remove kind parameter from backward. Add kind parameter to evaluate_mesh.
* evaluate_mesh does no longer take the mesh as input. Instead it is using parameter kind and N.
* Remove use_einsum from TensorProductSpace.evaluate.
* Remove method evaluate_derivative as only backward_primitive is currently in use.

## Breaking Changes

- [x] Yes (describe below)
- [ ] No

    T.backward(u, kind="uniform")

will no longer work. Use

    T.evaluate_mesh(u, kind="uniform") 

Furthermore, evaluate has one less parameter since use_einsum is removed.

## Checklist

- [x] Tests added or updated
- [x] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [x] Coverage does not decrease
- [x] Git history is clean (squash/fixup before merge)

### CODE REVIEW

File: .github/workflows/pytest.yml [hunk 1]
```diff
@@ -38,6 +38,8 @@ jobs:
         run: |
           uv run pytest --num-devices=2
           uv run pytest --float64 --num-devices=2
+          uv run pytest -m slow --num-devices=2
```
Comment (augustfe) [inline .github/workflows/pytest.yml:41]: The slow tests are run right below this?

File: src/jaxfun/galerkin/orthogonal.py [hunk 2]
```diff
@@ -176,34 +203,39 @@ def derivative_coeffs(self, c: Array, k: int = 0) -> Array:
...
-    ) -> Array:
-        """Implementation of backward transform."""
-        xj = self.mesh(kind=kind, N=N)
+    @jax.jit(static_argnums=(0, 2))
+    def backward(self, c: Array, N: int | None = None) -> Array:
+        """Return series evaluation at quadrature points (possibly with padding).
+
+        Args:
+            c: Coefficients of orthogonal series (length <= self.N).
+            N: Number of points. Must be >= self._num_quad_points. Defaults to
```
Comment (augustfe) [inline src/jaxfun/galerkin/orthogonal.py:212]: What's the difference between `self.N` and `self._num_quad_points`? That is an unintuitive difference if there is one. Also don't like that the private property `_num_quad_points` is exposed in the docstring, either make it public or don't expose it imo

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 3]
```diff
@@ -212,169 +212,105 @@ def map_expr_reference_domain(self, u: sp.Expr) -> sp.Expr:
             u = space.map_expr_reference_domain(u)
         return u
 
-    # Cannot jit_vmap since tensor product mesh.
-    @jax.jit(static_argnums=(0, 3))
     def evaluate_mesh(
-        self, x: list[Array], c: Array, use_einsum: bool = False
```
Comment (augustfe) [inline src/jaxfun/galerkin/tensorproductspace.py:218]: With this change, have we lost the ability to easily evaluate the function at arbitrary points? Are we now limited to just the meshes?

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 4]
```diff
@@ -212,169 +212,105 @@ def map_expr_reference_domain(self, u: sp.Expr) -> sp.Expr:
...
                 for i in range(dim)
             ]
 
-            # Build einsum string for arbitrary dim:
-            #   C[0]:"ja", C[1]:"jb", ..., c_loc:"ab..." → "j"
-            # Contracts all spectral axes; j is the evaluation-point axis.
-            dc = "abcdef"[:dim]
-            einsum_str = ",".join(f"j{ch}" for ch in dc) + f",{dc}->j"
+            cache_key = ("evaluate_spmd",)
+            if cache_key not in self._spmd_local_fn_cache:
```
Comment (augustfe) [inline src/jaxfun/galerkin/tensorproductspace.py:285]: Is it perhaps preferable to use the indices directly, instead of relying on the strings? (e.g. if there are more than 6 dimensions for some reason)

```python
>>> jnp.einsum('i,i->', x, y)  # explicit form
Array(16, dtype=int32)
>>> jnp.einsum(x, (0,), y, (0,))  # implicit form via indices
Array(16, dtype=int32)
>>> jnp.einsum(x, (0,), y, (0,), ())  # explicit form via indices
Array(16, dtype=int32)
```

File: src/jaxfun/galerkin/tensorproductspace.py [hunk 5]
```diff
@@ -383,6 +319,15 @@ def get_orthogonal(self) -> TensorProductSpace:
...
         )
 
+    def get_transposed_sharding(self, sharding: NamedSharding) -> NamedSharding:
+        """Return the sharding with unsharded and sharded axes transposed."""
+        if sharding is self._spectral_sharding:
+            return self._physical_sharding
+        elif sharding is self._physical_sharding:
+            return self._spectral_sharding
+        else:
+            raise ValueError("Provided sharding does not match spectral or physical.")
```
Comment (augustfe) [inline src/jaxfun/galerkin/tensorproductspace.py:329]:
```suggestion
            raise ValueError(f"Provided {sharding=} does not match spectral or physical.")
```

**Review by augustfe — COMMENTED** [review]
Looks good!

File: .github/workflows/pytest.yml (hunk 1 above)
Comment (mikaem) [reply .github/workflows/pytest.yml:41]: Just the one marked with spmd.

File: src/jaxfun/galerkin/orthogonal.py (hunk 2 above)
Comment (mikaem) [reply src/jaxfun/galerkin/orthogonal.py:212]: There's only a difference for BCGeneric spaces. These spaces have small N, but the same _num_quad_points as the orthogonal spaces they add to. Yes, should be num_quad_points in the docstring.

File: src/jaxfun/galerkin/tensorproductspace.py (hunk 3 above)
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:218]: No. Use `evaluate` for arbitrary points. `evaluate_mesh` only evaluates tensor product meshes, uniform or quadrature. With given N. So cannot use arbitrary tensor product meshes anymore.

File: src/jaxfun/galerkin/tensorproductspace.py (hunk 4 above)
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:285]: If you think there should be more than 6 dimensions, use "abcdefghijklmnopqrs#$%"[:dim]. I think current works.

File: src/jaxfun/galerkin/tensorproductspace.py (hunk 5 above)
Comment (mikaem) [reply src/jaxfun/galerkin/tensorproductspace.py:329]: Should there be an equal signe there? {sharding=}

**Comment by mikaem** [comment]
I'm putting fixes in next branch [sharding](https://github.com/spectralDNS/jaxfun/tree/sharding) @augustfe.
